File: CustomerSales/src/Entities/Customer.cs
namespace Api.Entities
{
    public record Customer
    {
        private CustomerStatusEnum _status;
        
        public Guid CustomerId { get; set; }

        public string Status
        {
            get => _status.ToString();
            set => _status = (CustomerStatusEnum)Enum.Parse(typeof(CustomerStatusEnum), value);
        }
        public DateTime UtcCreatedAt { get; set; }
        public string Name { get; set; }
        public string? Email { get; set; }
        // typically phone numbers are digits, but:
        // - it's possible someone could use the letters on the digits for their number?
        // - and / or they want to record a plus symbol (+)?
        // if something more complex than a string is needed, then the phone number should probably become a class.
        public string? PhoneNumber { get; set; }
        public List<SalesOpportunity> SalesOpportunities { get; set; }

        public Customer()
        {
            CustomerId = Guid.NewGuid();
            _status = CustomerStatusEnum.Unknown;
            UtcCreatedAt = DateTime.UtcNow;
            Name = string.Empty;
            SalesOpportunities = [];
        }
    }
}

File: frontend/src/Entities/Customer.ts
import SalesOpportunity from "./SalesOpportunity.ts";

export default class Customer {
    public customerId: string
    public status: string
    public utcCreatedAt: Date
    public name: string
    public email: string | null
    public phoneNumber: string | null
    public salesOpportunities: SalesOpportunity[] | null

    constructor(customerId: string, status: string, utcCreatedAt: Date, name: string, email: string | null, phoneNumber: string | null) {
        this.customerId = customerId
        // todo-at: some validators may be needed here to for example catch empty strings where inappropriate, etc.
        // - customer id can be empty
        // - status should probably have a default value (from the backend enum)
        // - name should have a value
        // - either email or phoneNumber should have a value
        // - sales opportunities can be set elsewhere
        this.status = status
        this.utcCreatedAt = utcCreatedAt
        this.name = name
        this.email = email
        this.phoneNumber = phoneNumber
        this.salesOpportunities = null
    }
}
